applications/matter_bridge/src/bridged_device_types/ble_onoff_light_data_provider.cpp

```diff
@@ -0,0 +1,180 @@
+					 dataSize);
+	}
+}
+
+void BleOnOffLightDataProvider::GattWriteCallback(struct bt_conn *conn, uint8_t err,
+						  struct bt_gatt_write_params *params)
+{
+	BLEBridgedDevice *dev = GetBLEConnectivityManager().FindBLEBridgedDevice(conn);
+
+	if (!dev || !params) {
+		return;
+	}
+
+	if (params->length != sizeof(mOnOff)) {
+		return;
+	}
+
+	if (!dev->provider) {
+		return;
+	}
+
+	BleOnOffLightDataProvider *provider = reinterpret_cast<BleOnOffLightDataProvider *>(dev->provider);
+
+	/* Save data received in GATT write response. */
+	memcpy(&provider->mOnOff, params->data, params->length);
+
+	DeviceLayer::PlatformMgr().ScheduleWork(NotifyAttributeChange, reinterpret_cast<intptr_t>(provider));
+}
+
+CHIP_ERROR BleOnOffLightDataProvider::UpdateState(chip::ClusterId clusterId, chip::AttributeId attributeId,
+						  uint8_t *buffer)
+{
+	if (clusterId != Clusters::OnOff::Id) {
+		return CHIP_ERROR_INVALID_ARGUMENT;
+	}
+
+	LOG_INF("Updating state of the BleOnOffLightDataProvider, cluster ID: %u, attribute ID: %u.", clusterId,
+		attributeId);
+
+	switch (attributeId) {
+	case Clusters::OnOff::Attributes::OnOff::Id: {
+		memcpy(mGattWriteDataBuffer, buffer, sizeof(mOnOff));
+
+		mGattWriteParams.data = mGattWriteDataBuffer;
+		mGattWriteParams.offset = 0;
+		mGattWriteParams.length = sizeof(mOnOff);
+		mGattWriteParams.handle = mLedCharacteristicHandle;
+		mGattWriteParams.func = BleOnOffLightDataProvider::GattWriteCallback;
+
+		int err = bt_gatt_write(mDevice->mConn, &mGattWriteParams);
+		if (err) {
+			LOG_ERR("GATT write operation failed");
+			return CHIP_ERROR_INTERNAL;
+		}
+
+		return CHIP_NO_ERROR;
+	}
+	default:
+		return CHIP_ERROR_INVALID_ARGUMENT;
+	}
+
+	return CHIP_NO_ERROR;
+}
+
+bt_uuid *BleOnOffLightDataProvider::GetServiceUuid()
+{
+	return sServiceUuid;
+}
+
+int BleOnOffLightDataProvider::ParseDiscoveredData(struct bt_gatt_dm *discoveredData)
+{
+	const struct bt_gatt_dm_attr *gatt_chrc;
+	const struct bt_gatt_dm_attr *gatt_desc;
+
+	gatt_chrc = bt_gatt_dm_char_by_uuid(discoveredData, sUuidLED);
+	if (!gatt_chrc) {
+		LOG_ERR("No LED characteristic found.");
+		return -EINVAL;
+	}
+
+	gatt_desc = bt_gatt_dm_desc_by_uuid(discoveredData, gatt_chrc, sUuidLED);
+	if (!gatt_desc) {
+		LOG_ERR("No LED characteristic value found.");
+		return -EINVAL;
+	}
+	mLedCharacteristicHandle = gatt_desc->handle;
+
+	gatt_chrc = bt_gatt_dm_char_by_uuid(discoveredData, sUuidButton);
+	if (!gatt_chrc) {
+		LOG_ERR("No Button characteristic found.");
+		return -EINVAL;
+	}
+
+	gatt_desc = bt_gatt_dm_desc_by_uuid(discoveredData, gatt_chrc, sUuidButton);
+	if (!gatt_desc) {
+		LOG_ERR("No Button characteristic value found.");
+		return -EINVAL;
+	}
+	mButtonCharacteristicHandle = gatt_desc->handle;
+
+	gatt_desc = bt_gatt_dm_desc_by_uuid(discoveredData, gatt_chrc, sUuidCcc);
+	if (!gatt_desc) {
+		LOG_ERR("No CCC descriptor found.");
+		return -EINVAL;
+	}
+	mCccHandle = gatt_desc->handle;
```

Comment:
Nit: It seems that it can be simplified somehow using for.

Reply:
Not quite, the CCC descriptor doesn get char, but only descriptor. So we could simplify only led and button part, so create for loop to iterate through 2 elements :)

Reply:
We could also create a list of handles, however I believe it would be harder to access them in other parts of the code, as every handle is used for other kind of interaction.

Reply:
Ok, let's leave it as it is :)